### DISCUSSION TITLE
Is there any way to install latest version of podman on Debian/Ubuntu/RaspberryPi ?

### DISCUSSION THREAD

**Discussion by gitsang** [containers/podman#24633] [Q&A] [answered]
I have tried almost all possible online tutorials, but I still can't install the latest podman on my Raspberry Pi 4B. I can only install up to version 4.3.1. I'm wondering if there is really a way to install >v4.7.0 or even v5.3.0?

**Comment by afbjorklund** [comment]
You can run it in a virtual machine (like on macOS or Windows), or you can build it from source (although instructions are outdated).

* https://github.com/podman-desktop/podman-desktop/discussions/5762

**Comment by MushiTheMoshi** [comment]
how to install it natively in debian:bookworm/ubuntu? I don't want to use a virtual machine, just want to install latest version...

**Reply by gitsang** [reply]
I finally found that the only way to install latest version of podman in debian is build it from source.

**Reply by MushiTheMoshi** [reply]
I tried that but it seems instructions have missing dependencies because when I build it I get errors with make...

**Reply by sbrivio-rh** [reply]
Two options:

- step up and help the maintainers of https://tracker.debian.org/pkg/libpod
- `apt-get install build-essential catatonit conmon crun go libgpgme-dev netavark systemd-dev && git clone https://github.com/containers/podman.git && cd podman && make`. Install as root with `make install`. It works well enough that we automate it at https://passt.top/passt/tree/test/Makefile?id=09478d55fe1a21f8c55902399df84d13867e71be#n77 (yes, that runs _also_ on Debian testing)
- step up and help the maintainers of https://tracker.debian.org/pkg/libpod

...yes, those are **two** options.

**Reply by afbjorklund** [reply]
Or use a build container, in the link above I used podman3 to build podman4 (and podman4 to build podman5)... Then you don't have to install the build tools on your host, only the requirements (libs)

https://github.com/podman-desktop/podman-desktop/discussions/5762#discussioncomment-8311142

**Reply by afbjorklund** [reply]
> * step up and help the maintainers of https://tracker.debian.org/pkg/libpod

Seems that it now lives at https://tracker.debian.org/pkg/podman (libpod is gone)

**Reply by MushiTheMoshi** [reply]
hey, thanks for the swift responses, appreciated. OK, build containers sound like a to go approach, can't find them though, which link are you referring to?

**Reply by afbjorklund** [reply]
There are a lot of mandatory config files as well, but I wrote down _most_ of it...

https://github.com/afbjorklund/podman-full (including a tarball, if in a hurry)

**Reply by MushiTheMoshi** [reply]
thanks, will take a look, greatly appreciated.

**Reply by afbjorklund** [reply]
The rust compiler ran out of memory on the Raspberry Pi, so had to throttle it to one GB per core

Cross-building (from amd64 to arm64) seems to be broken with the legacy builder, only BuildKit

```console
anders@hallon:~ $ apt-cache madison podman
    podman | 4.3.1+ds1-8+deb12u1 | http://deb.debian.org/debian bookworm/main arm64 Packages
anders@hallon:~ $ podman --version
podman version 4.9.3
```

Building on the Raspberry Pi takes about an hour or so (!), depending on the hardware model...

**Reply by sbrivio-rh** [reply]
> Seems that it now lives at https://tracker.debian.org/pkg/podman (libpod is gone)

I missed that. So now the latest package (in unstable and testing, of course, not stable) is 5.3.1, which is 1. quite impressive and 2. the latest Podman version available.

Why wouldn't you use that? ...and report/fix bugs if something is wrong, contributing back to the project (Debian, and to some extent Podman)?

**Reply by afbjorklund** [reply]
> Why wouldn't you use that?

If you are using Debian sid or Fedora, then the system packages would be new enough to not have to look to the vendor for packages (similar to docker.io vs docker-ce).

https://github.com/podman-desktop/podman-desktop/discussions/5762#discussioncomment-8311142

https://repology.org/project/podman

**Reply by sbrivio-rh** [reply]
> If you are using Debian sid

or Debian testing

> [...]

**Reply by MushiTheMoshi** [reply]
I am using debian 12/11, which is LTS... in which latest version of podman is not present...

**Reply by sbrivio-rh** [reply]
> I am using debian 12/11, which is LTS... in which latest version of podman is not present...

This is what I don't understand. You're using a distribution branch that **by design** has packages frozen for six months or longer (minus _some_ security fixes). Installing the latest version of Podman on it completely defies the purpose of using Debian stable.

I mean, of course, you're free to do that even if I don't understand why anybody would do that, but I'd argue that it doesn't fit with the title of this discussion.

"Install latest version of Podman on Debian" implies you're using at least Debian testing, in my eyes.

**Reply by MushiTheMoshi** [reply]
I get your point, certainly in terms of stability this make a lot of sense. In this case, the requirement is about to install latest version in the server for whatever reason (which escape to my  knowledge, maybe to take advantage of a new feature). In any case Podman documentation might pinpoint to a compatibility matrix like any other product where it is specified supported version in X distributions, which is either absent or somewhere in the documentation that I could not found or pay attention to.

**Reply by afbjorklund** [reply]
> "Install latest version of Podman on Debian" implies you're using at least Debian testing, in my eyes.

My interpretation of the question was that it was about Debian stable/Ubuntu LTS/Raspberry Pi OS...

**Reply by trya2l** [reply]
I said below that using nix was an option, but it may not suit every use case. As I've learn with recent experience, it doesn't do the same job as debian packages which go to the lengths of creating services and default configuration files.

So as @sbrivio-rh mentionned, using debian testing could make sense.

After all, I did install raspbian bookworkm on my RPi3 and ended up adding testing sources to apt, while keeping stable as the default release channel.

Here are the steps I followed:

_NB: I am by no mean expert in this matter, so I can't guarantee it'll work for you_

- Add testing sources to apt

```bash
# /etc/apt/sources.list

deb http://deb.debian.org/debian stable main contrib non-free non-free-firmware
deb http://deb.debian.org/debian-security/ stable-security main contrib non-free non-free-firmware
deb http://deb.debian.org/debian stable-updates main contrib non-free non-free-firmware

# Add testing sources
deb http://deb.debian.org/debian testing main contrib non-free non-free-firmware
deb http://deb.debian.org/debian-security/ testing-security main contrib non-free non-free-firmware
deb http://deb.debian.org/debian testing-updates main contrib non-free non-free-firmware
```

_NB: There seems to be a modern version of apt sources. I don't know much about that so I'll leave it aside. Maybe Raspbian will update that in the next version_:

```bash
pi@rpi-raspbian:~ $ sudo apt update --audit
Hit:1 http://deb.debian.org/debian stable InRelease                                                                                  
Hit:2 http://archive.raspberrypi.com/debian bookworm InRelease                                                                       
Hit:3 http://deb.debian.org/debian-security stable-security InRelease      
Get:4 http://deb.debian.org/debian stable-updates InRelease [55.4 kB]
Get:5 https://pkgs.tailscale.com/stable/debian bookworm InRelease
Get:6 http://deb.debian.org/debian testing InRelease [175 kB]
Hit:7 http://deb.debian.org/debian-security testing-security InRelease
Get:8 http://deb.debian.org/debian testing-updates InRelease [49.6 kB]
Fetched 287 kB in 2s (147 kB/s)    
9 packages can be upgraded. Run 'apt list --upgradable' to see them.
Warning: http://archive.raspberrypi.com/debian/dists/bookworm/InRelease: Policy will reject signature within a year, see --audit for details
Audit: http://archive.raspberrypi.com/debian/dists/bookworm/InRelease: Sub-process /usr/bin/sqv returned an error code (1), error message is:
   Signing key on CF8A1AF502A2AA2D763BAE7E82B129927FA3303E is not bound:
              No binding signature at time 2025-05-12T11:05:23Z
     because: Policy rejected non-revocation signature (PositiveCertification) requiring second pre-image resistance
     because: SHA1 is not considered secure since 2026-02-01T00:00:00Z
Audit: The sources.list(5) entry for 'http://deb.debian.org/debian' should be upgraded to deb822 .sources
Audit: Missing Signed-By in the sources.list(5) entry for 'http://deb.debian.org/debian'
Audit: The sources.list(5) entry for 'http://deb.debian.org/debian-security' should be upgraded to deb822 .sources
Audit: Missing Signed-By in the sources.list(5) entry for 'http://deb.debian.org/debian-security'
Audit: The sources.list(5) entry for 'http://deb.debian.org/debian' should be upgraded to deb822 .sources
Audit: Missing Signed-By in the sources.list(5) entry for 'http://deb.debian.org/debian'
Audit: The sources.list(5) entry for 'http://deb.debian.org/debian' should be upgraded to deb822 .sources
Audit: Missing Signed-By in the sources.list(5) entry for 'http://deb.debian.org/debian'
Audit: The sources.list(5) entry for 'http://deb.debian.org/debian-security' should be upgraded to deb822 .sources
Audit: Missing Signed-By in the sources.list(5) entry for 'http://deb.debian.org/debian-security'
Audit: The sources.list(5) entry for 'http://deb.debian.org/debian' should be upgraded to deb822 .sources
Audit: Missing Signed-By in the sources.list(5) entry for 'http://deb.debian.org/debian'
Audit: The sources.list(5) entry for 'http://archive.raspberrypi.com/debian' should be upgraded to deb822 .sources
Audit: Missing Signed-By in the sources.list(5) entry for 'http://archive.raspberrypi.com/debian'
Audit: The sources.list(5) entry for 'https://pkgs.tailscale.com/stable/debian' should be upgraded to deb822 .sources
Audit: Consider migrating all sources.list(5) entries to the deb822 .sources format
Audit: The deb822 .sources format supports both embedded as well as external OpenPGP keys
Audit: See apt-secure(8) for best practices in configuring repository signing.
```

```bash
Notice: Some sources can be modernized. Run 'apt modernize-sources' to do so.
```


- As described in `man apt_preferences`, I then set stable (I guess you can set `bookworm` if you decided to pin it) as the default release channel

```bash
# /etc/apt/apt.conf

APT::Default-Release "stable";
```

- Update sources, then select testing target version of podman, which is up to date:

```bash
# /etc/apt/preferences

sudo apt update
sudo apt install -t testing podman # podman-compose
```

You should now be all set. Hope that'll help.

**Comment by sbrivio-rh** [comment]
`apt-get install podman`

**Reply by sbrivio-rh** [reply]
Well, what can I say?

```
# apt-get install podman
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  libgpgme-dev libgpgme11t64 python3-gpg
The following packages will be upgraded:
  libgpgme-dev libgpgme11t64 podman python3-gpg
4 upgraded, 0 newly installed, 0 to remove and 934 not upgraded.
Need to get 25.4 MB of archives.
After this operation, 38.1 MB of additional disk space will be used.
Do you want to continue? [Y/n] 
Get:1 file:/etc/apt/mirrors/debian.list Mirrorlist [30 B]
Get:2 https://deb.debian.org/debian sid/main amd64 libgpgme-dev amd64 1.24.0-2 [524 kB]
Get:3 https://deb.debian.org/debian sid/main amd64 libgpgme11t64 amd64 1.24.0-2 [344 kB]
Get:4 https://deb.debian.org/debian sid/main amd64 podman amd64 5.3.1+ds1-7 [24.1 MB]
Get:5 https://deb.debian.org/debian sid/main amd64 python3-gpg amd64 1.24.0-2 [412 kB]
Fetched 25.4 MB in 0s (62.8 MB/s)
Reading changelogs... Done
(Reading database ... 127395 files and directories currently installed.)
Preparing to unpack .../libgpgme-dev_1.24.0-2_amd64.deb ...
Unpacking libgpgme-dev (1.24.0-2) over (1.18.0-6+b1) ...
Preparing to unpack .../libgpgme11t64_1.24.0-2_amd64.deb ...
Unpacking libgpgme11t64:amd64 (1.24.0-2) over (1.18.0-6+b1) ...
Preparing to unpack .../podman_5.3.1+ds1-7_amd64.deb ...
Unpacking podman (5.3.1+ds1-7) over (5.2.1+ds1-2) ...
Preparing to unpack .../python3-gpg_1.24.0-2_amd64.deb ...
Unpacking python3-gpg (1.24.0-2) over (1.18.0-6+b1) ...
Setting up libgpgme11t64:amd64 (1.24.0-2) ...
Setting up podman (5.3.1+ds1-7) ...
Removing obsolete conffile /etc/profile.d/podman-docker.csh ...
Removing obsolete conffile /etc/profile.d/podman-docker.sh ...
Setting up python3-gpg (1.24.0-2) ...
Setting up libgpgme-dev (1.24.0-2) ...
Processing triggers for man-db (2.11.2-3) ...
Processing triggers for libc-bin (2.40-3) ...
Scanning processes...                                                                                                                                                                                                                           
Scanning linux images...                                                                                                                                                                                                                        

Running kernel seems to be up-to-date.

No services need to be restarted.

No containers need to be restarted.

No user sessions are running outdated binaries.

No VM guests are running outdated hypervisor (qemu) binaries on this host.
# podman --version
podman version 5.3.1
```

I think Debian maintainers are doing an amazing job with this.

**Reply by afbjorklund** [reply]
> I think Debian maintainers are doing an amazing job with this.

No argument there, especially considering it is only one person.

**Reply by gitsang** [reply]
Can you provide more specific information? For example, the repository used, OS version, etc.?

I have try in my raspberry pi 4B（"Debian GNU/Linux 12 (bookworm)"），it still v4.3.1。

**Reply by sbrivio-rh** [reply]
> Can you provide more specific information? For example, the repository used, OS version, etc.?

I have both Debian testing ("trixie") and Debian unstable ("sid") running, with default repositories for those versions.

> I have try in my raspberry pi 4B（"Debian GNU/Linux 12 (bookworm)"），it still v4.3.1。

I don't think it makes much sense to run Debian stable on a Raspberry Pi and then try to install the latest version of something, especially Podman. You install Debian stable if you want a extremely low, system-wide risk of regressions.

This conflicts with the fact that you're then trying to install the latest version of Podman. **I would suggest that you switch to Debian testing** (or unstable), which are actually quite stable. See also https://www.debian.org/doc/manuals/debian-faq/choosing.en.html#s3.1.3 / https://www.debian.org/doc/manuals/debian-faq/choosing.zh-cn.html#s3.1.3.

**Comment by trya2l** [comment]
You can also use [nix](https://nix.dev/install-nix) on debian. Personally I use [home-manager](https://nix-community.github.io/home-manager/index.xhtml#sec-install-standalone). You just have to add it to your `home.packages`

```nix
#home.nix
#...
packages = (with pkgs; [
   podman
   podman-compose
   #docker-compose
]
#...
```

```sh
podman --version
podman version 5.2.3
```

**Reply by ciolben** [reply, marked as answer]
What's needed is only:
`$ curl -L https://nixos.org/nix/install | sh -s -- --daemon`
then open a new terminal and type:
`$ nix-env -iA nixpkgs.podman`

You can verify by typing:
```
$ podman --version
podman version 5.4.1
```

(you may need to also install `sudo apt install uidmap`)

**Comment by andrewchen5678** [comment]
I just vibe coded scripts to build the latest podman 5.8.0 on docker and export the built packages for debian 13 trixie and ubuntu 24.04 noble https://github.com/andrewtheguy/podman-package, so far I have only tested rootless mode with just the podman binary with systemd quadlets. Looks like it is not difficult to make the latest podman version packages for debian based distros at least, I just hope that the podman team can maintain package repo with the latest version for multiple mainstream distros like docker because lack of the ability to install the latest version easily on mainstream distros really hinders podman's adaption.
